### PR TITLE
remove old callbacks in Request.JSONP.request_map

### LINKED ISSUE #1293: Methods not removed from the request map of JSONP
It isn't a serious issue for a single request but it leads to needless accumulation of functions when many requests are being sent.
http://jsfiddle.net/L576cbyn/

If the following line is put in the *clear* method of *Request.JSONP* class the problem is solved.
```javascript
delete Request.JSONP.request_map['request_' + this.index];
```
I plan to pull that and a small refactoring of the *send* method.

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/6600759-methods-not-removed-from-the-request-map-of-jsonp?utm_campaign=plugin&utm_content=tracker%2F22069&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F22069&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by SergioCrisostomo**
fixes #1293

Add option to remove old `Request.JSONP.request_map` callbacks. Defaults to `false` to be BC.

### CODE REVIEW

**Comment by DimitarChristoff** [comment]
since that's not official API (the stored responses) BC should be ignored
and data should be removed by default. anyone who's augmented it to use
request_map to cache results clearly know what they are doing and can
adjust their code accordingly imo.

On Tuesday, August 11, 2015, Sergio Crisostomo <notifications@github.com>
wrote:

> fixes #1293 <https://github.com/mootools/mootools-more/issues/1293>
>
> Add option to remove old Request.JSONP.request_map callbacks. Defaults to
> false to be BC.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/mootools/mootools-more/pull/1315
> Commit Summary
>
>    - remove old callbacks in Request.JSONP.request_map
>
> File Changes
>
>    - *M* Docs/Request/Request.JSONP.md
>    <https://github.com/mootools/mootools-more/pull/1315/files#diff-0> (1)
>    - *M* Source/Request/Request.JSONP.js
>    <https://github.com/mootools/mootools-more/pull/1315/files#diff-1>
>    (19)
>    - *M* Specs/Request/Request.JSONP.js
>    <https://github.com/mootools/mootools-more/pull/1315/files#diff-2>
>    (28)
>    - *M* Tests/Specs/assets/jsonp.js
>    <https://github.com/mootools/mootools-more/pull/1315/files#diff-3> (4)
>
> Patch Links:
>
>    - https://github.com/mootools/mootools-more/pull/1315.patch
>    - https://github.com/mootools/mootools-more/pull/1315.diff
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mootools/mootools-more/pull/1315>.
>


-- 
Dimitar Christoff

"JavaScript is to JAVA what hamster is to ham"
@D_mitar - https://github.com/DimitarChristoff

**Comment by arian** [comment]
Agreed with @DimitarChristoff

**Comment by SergioCrisostomo** [comment]
Nice feedback! will update soonish and ping back.

**Comment by SergioCrisostomo** [comment]
@arian @DimitarChristoff  updated

File: Source/Request/Request.JSONP.js [hunk 1]
```diff
@@ -119,19 +119,20 @@ Request.JSONP = new Class({
 		return !!this.running;
 	},
 
-	clear: function(){
+	clear: function(index){
```
Comment (anutron) [inline Source/Request/Request.JSONP.js]: Hmmm.

What if the index isn't passed in? If it's required, that's a breaking change.

What if the index passed in isn't the running request?

Comment (DimitarChristoff) [inline Source/Request/Request.JSONP.js]: it's not breaking because worst case scenario, it will just try to delete an undefined property. 

I have a better fix which does not change the API - storing the index within the script tag, then auto deleting

http://jsfiddle.net/dimitar/frxnqj5q/1/

File: Specs/Request/Request.JSONP.js [hunk 2]
```diff
@@ -52,4 +57,29 @@ describe('Request.JSONP', function(){
 
 	});
 
+	it('should clear the request callback map', function(){
```
Comment (DimitarChristoff) [inline Specs/Request/Request.JSONP.js]: why don't we use `function(done){ ... done() ... }` for async but we have `waitsFor(<ambiguous number>, fn);`

File: Source/Request/Request.JSONP.js (hunk 1 above)
Comment (anutron) [inline Source/Request/Request.JSONP.js]: Good stuff. Might suggest we used data-jsonp-index or something rather than the element attribute, but whatever, it doesn't matter really.

Comment (DimitarChristoff) [inline Source/Request/Request.JSONP.js]: i considered it but it's so short lived - initially even thought of storage but that meant storage gc when element is removed. i'd agree about `data-prop` as it ought to be slightly cheaper.

Comment (DimitarChristoff) [inline Source/Request/Request.JSONP.js]: easier fix still. one liner. 
```
    Request.JSONP.request_map['request_' + index] = function(){
        delete Request.JSONP.request_map['request_' + index];
        this.success(arguments, index);
    }.bind(this);
```
so
`+ delete Request.JSONP.request_map['request_' + index];`

File: Specs/Request/Request.JSONP.js (hunk 2 above)
Comment (SergioCrisostomo) [inline Specs/Request/Request.JSONP.js]: @DimitarChristoff: we are still using Jasmine `1.3` and the `done()` API is from `2.0`. We can do that upgrade in another PR, that would be good.

**Comment by SergioCrisostomo** [comment]
Updated again. No `index` anymore and using Dimitar's _oneliner_, much better idea.

**Comment by DimitarChristoff** [comment]
@SergioCrisostomo check https://github.com/mootools/mootools-more/compare/master...DimitarChristoff:SergioCrisostomo-fix1293 - missed an index and such.

**Comment by SergioCrisostomo** [comment]
Nice! Yeah, that index was already there from before and is unused. Will cherry pick your work and use it. thanks!

**Comment by bhoomi01** [comment]
This is very great and brilliant information.
[gapps download apk](http://gappsapk.co/gapps-for-pc/)

**Comment by promocode321** [comment]
Get idea this question about [Retrosuperfuture promo codes](https://www.promoocodes.com/) It is my key period I reach here. The procedure you generate a blog webpage especially raised and head blowing.

**Comment by Oliviajones77** [comment]
what a nice article. thank you so much for this gode articel ...

[norton.com/setup](https://manage-notron.com/) |[ mcafee.com/activate](http://how-to-activate.net/) | [office.com/setup](https://offiice-products.com/)

**Comment by crazyloveboy** [comment]
good post 
https://www.jiotvforpcapkdownload.in/

**Comment by techproofs** [comment]
The matter is good and sufficient . The writing skills are awesome.Just take a look on this also 
<a href="https://www.techproofs.com/free-netflix-cookies-2019/">Free Netflix Cookies</a>
<a href="https://www.techproofs.com/free-netflix-accounts-and-passwords-genrator-2019/">Netflix Cracked Account</a>

**Comment by techproofs** [comment]
The matter is good and sufficient. The writing skills are awesome.Just take a look on this also 
 [How to install WhatsApp on pc windows 7](https://www.techproofs.com/install-whatsapp-in-pc/l)
